StandardLibrary/src/GetOpt.dfy

```diff
@@ -0,0 +1,435 @@
+      x + GetHelp2(opts[1..], longLen)
+  }
+
+  function GetCmdHelp(opts : seq<Param>, commandLen : nat) : string
+  {
+    if |opts| == 0 then
+      ""
+    else
+      var x :=
+        if opts[0].Command? then
+          GetCommandHelp(opts[0], commandLen)
+        else
+          "";
+      x + GetCmdHelp(opts[1..], commandLen)
+  }
+
+  function GetLongLen(opts : seq<Param>, max : nat := 0) : nat
+  {
+    if |opts| == 0 then
+      max
+    else
+      var x := |GetLongHelp(opts[0], 0)|;
+      var newMax := if x > max then x else max;
+      GetLongLen(opts[1..], newMax)
+  }
+
+  function GetCommandLen(opts : seq<Param>, max : nat := 0) : nat
+  {
+    if |opts| == 0 then
+      max
+    else
+      var x := if opts[0].Command? then |opts[0].options.name| else 0;
+      var newMax := if x > max then x else max;
+      GetCommandLen(opts[1..], newMax)
+  }
+
+  function {:tailrecursion} GetMaps(
+    opts : seq<Param>,
+    longMap : map<string, bool> := map[],
+    shortMap : map<char, string> := map[],
+    commandMap : CommandMap := map[])
+    : (ret : Result<(map<string, bool>, map<char, string>, CommandMap), string>)
+    requires forall x <- shortMap :: shortMap[x] in longMap
+    ensures ret.Success? ==> (forall x <- ret.value.1 :: ret.value.1[x] in ret.value.0)
+  {
+    if |opts| == 0 then
+      Success((longMap, shortMap, commandMap))
+    else
+      var opt := opts[0];
+      if opt.Command? then
+        :- Need(opt.options.name !in commandMap, "Duplicate command in options : " + opt.options.name);
+        GetMaps(opts[1..], longMap, shortMap, commandMap[opt.options.name := opt.options])
+      else
+        :- Need(opt.name !in longMap, "Duplicate long name in options : " + opt.name);
+        var longMap := longMap[opt.name := opt.Opt?];
+        if opt.short.Some? then
+          var short := opt.short.value;
+          if short in shortMap then // can't be a `Need` because shortMap[short] required in message
+            Failure("Duplicate short char in options : '" + [short] + "' for " + opt.name + " and " + shortMap[short])
+          else
+            GetMaps(opts[1..], longMap[opt.name := opt.Opt?], shortMap[short := opt.name], commandMap)
+        else
+          GetMaps(opts[1..], longMap[opt.name := opt.Opt?], shortMap, commandMap)
+  }
+
+  function printFromFunction<T>(x: T): () {
+    ()
+  } by method {
+    print x,"\n";
+    return ();
+  }
+
+  function GetOptions(opts : Options, args : seq<string>)
+    : Result<Parsed, string>
+    requires 0 < |args|
+    decreases args
+  {
+    var (longMap, shortMap, commandMap) :- GetMaps(opts.params, map["help" := false]);
+    var context := Context(longMap, shortMap, commandMap, args[0]);
+    GetOptions2(args[1..], context)
+  }
+
+  datatype Context = Context (
+    longMap : map<string, bool>,
+    shortMap : map<char, string>,
+    commands : CommandMap,
+    command : string
+  )
+
+  function /*{:tailrecursion}*/ GetOptions2(
```

Comment:
What is the diff in the 2?

Reply:
GetOptions is the thing the user calls.  GetOptions2 is the recursive thing that implements most of GetOptions.